src/proto.c

```diff
@@ -584,15 +593,69 @@ bool answer_authreq(PgSocket *server, PktHdr *pkt)
 
 bool send_startup_packet(PgSocket *server)
 {
-	PgDatabase *db = server->pool->db;
-	const char *username = server->pool->user->name;
-	PktBuf *pkt;
-
-	pkt = pktbuf_temp();
-	pktbuf_write_StartupMessage(pkt, username,
-				    db->startup_params->buf,
-				    db->startup_params->write_pos);
-	return pktbuf_send_immediate(pkt, server);
+	PgPool *pool = server->pool;
+	PgDatabase *db = pool->db;
+	const char *username = pool->user->name;
+	PktBuf *pkt = pktbuf_temp();
+	PgSocket *client = NULL;
+
+	pktbuf_start_packet(pkt, PKT_STARTUP);
+	pktbuf_put_bytes(pkt, db->startup_params->buf, db->startup_params->write_pos);
+
+	/*
+	 * If the next client in the list is a replication connection, we need
+	 * to do some special stuff for it.
+	 */
+	client = first_socket(&pool->waiting_client_list);
+	if (client && client->replication) {
+		server->replication = client->replication;
+		pktbuf_put_string(pkt, "replication");
+		slog_debug(server, "send_startup_packet: creating replication connection");
+		pktbuf_put_string(pkt, replication_type_parameters[server->replication]);
+
+		/*
+		 * For a replication connection we apply the varcache in the
+		 * startup instead of through SET commands after connecting.
+		 * The main reason to do so is because physical replication
+		 * connections don't allow SET commands. A second reason is
+		 * because it allows us to skip running the SET logic
+		 * completely, which normally requires waiting on multiple
+		 * server responses. This SET logic is normally executed in the
+		 * codepath where we link the client to the server
+		 * (find_server), but because we link the client here already
+		 * we don't run that code for replication connections. Adding
+		 * the varcache parameters to the startup message allows us to
+		 * skip the dance that involves sending Query packets and
+		 * waiting for responses.
+		 */
```

Comment:
Could we also apply this to normal connections, to get the same optimization benefit?

Reply:
Then we would need to link normal connections here already too. We don't do that currently. I'd expect that to cause some problems if we did that. It might be possible, but I don't think it fits well in this PR to do so.